dashboard/lib/feature_mode_manager.rb

```diff
@@ -0,0 +1,130 @@
+# A class that defines "feature modes" for Gatekeeper and DCDO configuration. There are currently
+# three modes defined. The "normal" mode is intended for use throughout most of the year, the
+# "scale" mode is intended for use during the hour of code (when we peak at 10x normal traffic),
+# and the "emergency" mode is for use during severe database overload.
+#
+# Below are the groupings of features that are enabled at each scale level:
+#
+# normal:
+#     Enable all features.
+# scale:
+#     Save progress only on sharable levels for HOC tutorials.
+#     Disable puzzle rating and hints for all tutorials.
+#     Log hoc activity with 10% sampling.
+#     Extend cache lifetimes for HOC pages
+# emergency:
+#     Same as Scale, plus disable sharing for HOC tutorials.
+
+require 'cdo/hip_chat'
+
+class FeatureModeManager
+
+  # A map from mode symbol to gatekeeper and dcdo settings. The gatekeeper settings are
+  # broken down into gatekeeper_general_settings (which apply to all tutorials) and
+  # gatekeeper_hoc_tutorial_settings (which apply to high-scale HOC tutorials only).
+  MODE_SETTINGS_MAP = {
+    normal: {
+        gatekeeper_general_settings: {
+            puzzle_rating: true,
+            hint_view_request: true
+        },
+        gatekeeper_hoc_tutorial_settings: {
+          postMilestone: true,
+          shareEnabled: true,
+        },
+        dcdo_settings: {
+          hoc_activity_sample_weight: 1,
+          public_proxy_max_age: 3.minutes.to_i,
+          public_max_age: 6.minutes.to_i,
+        }
+    },
+    scale: {
+      gatekeeper_general_settings: {
+          puzzle_rating: false,
+          hint_view_request: false
+      },
+      gatekeeper_hoc_tutorial_settings: {
+        postMilestone: false,
+        shareEnabled: true,
+      },
+      dcdo_settings: {
+        hoc_activity_sample_weight: 10,
+        public_proxy_max_age: 4.hours.to_i,
+        public_max_age: 8.hours.to_i,
+      }
+    },
+    emergency: {
+      gatekeeper_general_settings: {
+          puzzle_rating: false,
+          hint_view_request: false
+      },
+      gatekeeper_hoc_tutorial_settings: {
+        postMilestone: false,
+        shareEnabled: false,
+      },
+      dcdo_settings: {
+        hoc_activity_sample_weight: 10,
+        public_proxy_max_age: 24.hours.to_i,
+        public_max_age: 48.hours.to_i,
+      }
+    }
+  }.with_indifferent_access
+
+  MODES = MODE_SETTINGS_MAP.keys
+
+  # Updates the feature mode to the given mode.
+  def self.set_mode(mode, gatekeeper, dcdo, hoc_script_names)
```

Comment:
The manager is invoked by feature_mode_controller, which will be submitted in my next PR.  

The UI of the controller is described by this mockup:
https://docs.google.com/presentation/d/1rdIOXcrAlpKKxiP7k_cSfUCb38rbbaTncye_vXZPx-Y/edit